File: 3rd_Year/ca341/Assigment/euler/problems/euler-131-prime-cube-partnership.go/euler-131-prime-cube-partnership.go
package main

/*
Import needed packages for problem at hand
fmt imported for printing
os is used for getting command line input
strconv used for converting strings to ints
*/
import (
	"fmt"
	"os"
	"strconv"
)

/*
Function that takes an integer as an argument and returns
true if the integer is prime
*/
func isPrime(num int) bool {
	if num <= 1 {
		return false
	} else if num <= 3 {
		return true
	} else if num%2 == 0 || num%3 == 0 {
		return false
	} else {
		for i := 5; i*i <= num; i += 6 {
			if num%i == 0 || num%(i+2) == 0 {
				return false
			}
		}
	}
	return true
}

func main() {
	//using os to get command line input
	limit, _ := strconv.Atoi(os.Args[1])
	numP := 0
	var num int
	//for loop to count the numbers that satisfy the condition
	//of the problem of prime cube partnership
	for a := 0; num <= limit; a++ {
		//shortcut calculation found called cuban primes
		num = 3*a*(1+a) + 1
		if isPrime(num){
			numP++
		}
	}
	fmt.Printf("%d\n", numP)
}

//https://en.wikipedia.org/wiki/Cuban_prime

File: 3rd_Year/ca341/Assigment/euler/problems/euler-131-prime-cube-partnership.kt/euler-131-prime-cube-partnership.kt
/*
class Problem131 created consisting of the 
isPrime() class method to check for a prime number and
cubanPrime() class method to check for cuban prime numbers
*/
class Problem131{
    /*
    class method that takes one agument of type int
    and checks if that number is prime returning true
    if it is and false if not
    */
    fun isPrime(num: Int): Boolean {
        if (num <= 1) {
            return false
        } else if (num <= 3) {
            return true
        } else if (num % 2 == 0 || num % 3 == 0) {
            return false
        } else {
            var i = 5
            while (i * i <= num) {
                if (num % i == 0 || num % (i + 2) == 0) {
                    return false
                }
                i += 6
            }
        }
        return true
    }

    /*
    method that takes one integer and returns
    the generated cumban primes in the range provided
    */
    fun cubanPrimes(n: Int): Int{
        var numP = 0
        var num = 0
        var a = 0

        /*
        while loop that while less than the inputted number
        a count is create of all the cuban primes up to that number
        */
        while (num <= n) {
            // shortcut calculation found from cuban primes
            num = 3 * a * (1 + a) + 1
            if (isPrime(num)) {
                numP++
            }
            a++
        }
        return numP
    }
    

}

fun main(args: Array<String>) {
    //set the limit as the argument passed from command line
    val limit = args[0].toInt()
    val problem = Problem131()
    val answer = problem.cubanPrimes(limit)
    println(answer)
}
